File: client/game/city.lua
local City = {}

local dume = require("dume")
local Vector = require("brinevector")
local style = require("ui/style")

function City:new()
    local o = {}
    setmetatable(o, self)
    self.__index = self
    return o
end

-- Updates the city with data received from the server
-- in an UpdateCity packet.
function City:updateData(data, game)
    if self.buildTask ~= nil and data.buildTask == nil then
        self.previousBuildTask = self.buildTask
        game.eventBus:trigger("buildTaskCompleted", {
            city = self,
            buildTask = self.buildTask,
        })
    end

    if data.buildTask == nil then
        self.buildTask = nil
    end

    local previousOwner = self.owner

    for k, v in pairs(data) do
        self[k] = v
    end

    self.owner = game.players[self.ownerID]
    if self.owner == nil then print("city '" .. self.name .. "' has invalid owner!") end

    if self.owner ~= previousOwner then
        self.previousBuildTask = nil
    end

    self.buildings = {}
    for _, buildingName in ipairs(data.buildingNames) do
        local building = registry.buildings[buildingName]
        if building == nil then print("received invalid building '" .. buildingName .. "'!") end
        table.insert(self.buildings, building)
    end

    self.populationText = cv:parseTextMarkup("@size{14}{@color{rgb(0,0,0)}{%pop}}", style.default.text.defaultTextStyle, {pop=tostring(self.population)})
    self.populationParagraph = cv:createParagraph(self.populationText, {
        alignH = dume.Align.Center,
        alignV = dume.Align.Start,
        baseline = dume.Baseline.Middle,
        lineBreaks = false,
        maxDimensions = Vector(20, math.huge)
    })

    self.cityNameText = cv:parseTextMarkup("@size{10}{@color{rgb(255,255,255)}{%name}}", style.default.text.defaultTextStyle, {name=self.name})
    self.cityNameParagraph = cv:createParagraph(self.cityNameText, {
        alignH = dume.Align.Center,
        alignV = dume.Align.Start,
        baseline = dume.Baseline.Middle,
        lineBreaks = false,
        maxDimensions = Vector(100, math.huge)
    })

    self.cultureDefenseText = cv:parseTextMarkup("@size{12}{@color{rgb(255,255,255)}{+%bonus%percent}}", style.default.text.defaultTextStyle, {
        bonus = tostring(self.cultureDefenseBonus),
        percent = "%"
    })
    self.cultureDefenseParagraph = cv:createParagraph(self.cultureDefenseText, {
        alignH = dume.Align.Start,
        alignV = dume.Align.Start,
        baseline = dume.Baseline.Top,
        lineBreaks = false,
        maxDimensions = Vector(100, math.huge)
    })

    if self.buildTask ~= nil and self.buildTask.kind.building ~= nil then
        local c = string.sub(self.buildTask.kind.building.buildingName, 1, 1)
        local text = cv:parseTextMarkup("@bold{@size{18}{@color{rgb(0, 0, 0)}{%c}}}", style.default.text.defaultTextStyle, {c=c})
        self.buildTaskParagraph = cv:createParagraph(text, {
            alignH = dume.Align.Center,
            alignV = dume.Align.Center,
            baseline = dume.Baseline.Top,
            lineBreaks = false,
            maxDimensions = Vector(20, 20)
        })
    end

    self.happiness = 0
    for _, entry in ipairs(self.happinessSources) do self.happiness = self.happiness + entry.count end
    self.unhappiness = 0
    for _, entry in ipairs(self.unhappinessSources) do self.unhappiness = self.unhappiness + entry.count end
    self.health = 0
    for _, entry in ipairs(self.healthSources) do self.health = self.health + entry.count end
    self.sickness = 0
    for _, entry in ipairs(self.sicknessSources) do self.sickness = self.sickness + entry.count end
end

function City:estimateTurnsToBuild(buildTask)
    return math.ceil(
            (buildTask.cost - buildTask.progress)
            / self.yield.hammers
    )
end

function City:isTileWorked(tilePos)
    for _, workedTile in ipairs(self.workedTiles) do
        if workedTile.x == tilePos.x and workedTile.y == tilePos.y then
            return true
        end
    end
    return false
end

local function drawFivePointStar(cv, center, outerRadius, innerRadius)
    local angleStep = math.pi * 2 / 5

    for i=1,5 do
        local outerTheta = angleStep * (i - 1) - math.pi / 2
        local innerTheta = angleStep * (i - 1 / 2) - math.pi / 2

        local outerPos = Vector(math.cos(outerTheta), math.sin(outerTheta)) * outerRadius + center
        local innerPos = Vector(math.cos(innerTheta), math.sin(innerTheta)) * innerRadius + center

        if i == 1 then
            cv:moveTo(outerPos)
        else
            cv:lineTo(outerPos)
        end
        cv:lineTo(innerPos)
    end

    -- close the path
    cv:lineTo(center - Vector(0, outerRadius))
end

local numHouses = 3
local housePositions = {
    Vector(20, 25),
    Vector(50, 25),
    Vector(25, 30),
}
local houseScales = {
    25,
    25,
    55,
}

function City:renderHouses(cv)
    for i=1,numHouses do
        local housePos = housePositions[i]
        local houseScale = houseScales[i]

        cv:drawSprite("icon/house", housePos, houseScale / 1.424)
    end
end

local bubbleColorA = dume.rgb(61, 61, 62, 180)
local bubbleColorB = dume.rgb(40, 40, 41, 180)
local populationCircleColor = dume.rgb(182, 207, 174)
local black = dume.rgb(0, 0, 0)
local buildCircleColor = dume.rgb(244, 195, 204)

local function renderProgressBar(cv, pos, size, progress, projectedProgress, progressColor, projectedProgressColor)
    projectedProgress = math.clamp(projectedProgress, 0, 1)

    cv:beginPath()
    cv:rect(pos, Vector(size.x * progress, size.y))
    cv:solidColor(progressColor)
    cv:fill()

    cv:beginPath()
    cv:rect(pos + Vector(size.x * progress, 0), Vector(size.x * (projectedProgress - progress), size.y))
    cv:solidColor(projectedProgressColor)
    cv:fill()
end

function City:renderBubble(cv, game)
    -- Rounded rectangle (bubble background)
    local bubbleWidth = 100
    local bubbleHeight = 20
    cv:beginPath()
    cv:roundedRect(Vector(0, 70), Vector(bubbleWidth, bubbleHeight), 5)
    cv:linearGradient(Vector(0, 70), Vector(0, 90), bubbleColorA, bubbleColorB)
    cv:fill()

    if self.owner == game.thePlayer then
        -- Production progress bar
        if self.buildTask ~= nil then
            local progress = self.buildTask.progress / self.buildTask.cost
            local projectedProgress = (self.buildTask.progress + self.yield.hammers) / self.buildTask.cost
            renderProgressBar(cv, Vector(0, 80), Vector(bubbleWidth, bubbleHeight / 2), progress, projectedProgress,
                style.default.productionProgressBar.progressColor, style.default.productionProgressBar.positivePredictedProgressColor)
        end

        -- Population growth population bar
        local progress = self.storedFood / self.foodNeededForGrowth
        local projectedProgress = (self.storedFood + self.yield.food - self.consumedFood) / self.foodNeededForGrowth
        renderProgressBar(cv, Vector(0, 70), Vector(bubbleWidth, bubbleHeight / 2), progress, projectedProgress,
            style.default.populationProgressBar.progressColor, style.default.populationProgressBar.positivePredictedProgressColor)
    end

    -- Left circle, or five-point star if this is the capital
    local radius = 10
    local center = Vector(radius - 5, radius + 70)

    cv:beginPath()
    if self.isCapital then
        drawFivePointStar(cv, center, 18, 8)
    else
        cv:circle(center, radius)
    end
    cv:solidColor(populationCircleColor)
    cv:fill()
    cv:solidColor(black)
    cv:strokeWidth(1.5)
    cv:stroke()

    if self.owner == game.thePlayer then
        -- Right circle
        cv:beginPath()
        cv:circle(Vector(-radius + 5 + bubbleWidth, radius + 70), radius)
        cv:solidColor(buildCircleColor)
        cv:fill()
        cv:solidColor(black)
        cv:strokeWidth(1.5)
        cv:stroke()
    end

    -- Left circle text (population)
    cv:drawParagraph(self.populationParagraph, Vector(-5, 80))

    if self.owner == game.thePlayer then
        -- Right circle overlay, depending on the build task:
        -- * unit - unit head icon
        -- * building - first letter of building name (TODO: we should have icons for these)
        if self.buildTask ~= nil then
            if self.buildTask.kind.unit ~= nil then
                cv:drawSprite("icon/unit_head/" .. self.buildTask.kind.unit.unitKindID, Vector(-radius * 2 + 5 + bubbleWidth, 70), radius * 2)
            else
                cv:drawParagraph(self.buildTaskParagraph, Vector(-radius * 2 + 5 + bubbleWidth, 70))
            end
        end
    end

    -- City name
    cv:drawParagraph(self.cityNameParagraph, Vector(0, 80))

    if self.owner == game.thePlayer then
        -- Status bar
        local statusOffset = Vector(10, 55)

        local happyIcon
        if self.happiness >= self.unhappiness then happyIcon = "icon/happy"
        else happyIcon = "icon/unhappy" end
        cv:drawSprite(happyIcon, statusOffset, 15)

        statusOffset.x = statusOffset.x + 20

        local healthIcon
        if self.health >= self.sickness then healthIcon = "icon/health"
        else healthIcon = "icon/sick" end
        cv:drawSprite(healthIcon, statusOffset, 13)

        statusOffset.x = statusOffset.x + 20
        if self.cultureDefenseBonus ~= 0 then
            cv:drawParagraph(self.cultureDefenseParagraph, statusOffset)
        end
    end
end

function City:renderTradeDebugOverlay(cv, game)
    local network = game:getTradeNetworkAtPos(self.pos)
    if network ~= nil then
        local color = generatePastelColor(network.id)
        cv:solidColor(color)
        cv:beginPath()
        cv:circle(Vector(50, 50), 50)
        cv:strokeWidth(5)
        cv:stroke()
    end
end

function City:render(cv, game)
    self:renderHouses(cv)
    self:renderBubble(cv, game)
    if game.tradeDebugMode then
        self:renderTradeDebugOverlay(cv, game)
    end
end

return City

File: client/src/game/city.rs
use std::convert::TryInto;

use anyhow::{anyhow, bail};
use glam::UVec2;
use protocol::{HappinessEntry, HealthEntry, SicknessEntry, UnhappinessEntry};

use crate::{
    assets::Handle,
    registry::{Building, RegistryItemNotFound, Resource, UnitKind},
};

use super::{CityId, Culture, Game, PlayerId, Yield};

#[derive(Debug)]
pub struct City {
    data: protocol::UpdateCity,
    id: CityId,

    pos: UVec2,
    owner: PlayerId,
    city_yield: Yield,
    build_task: Option<BuildTask>,
    previous_build_task: Option<PreviousBuildTask>,
    buildings: Vec<Handle<Building>>,
    resources: Vec<Handle<Resource>>,
    culture: Culture,
}

impl City {
    pub fn from_data(data: protocol::UpdateCity, id: CityId, game: &Game) -> anyhow::Result<Self> {
        let mut city = Self {
            data: protocol::UpdateCity::default(),
            id,

            pos: Default::default(),
            owner: Default::default(),
            city_yield: Default::default(),
            build_task: None,
            previous_build_task: None,
            buildings: Vec::new(),
            resources: Vec::new(),
            culture: Culture::new(),
        };

        city.update_data(data, game)?;

        Ok(city)
    }

    pub fn update_data(&mut self, data: protocol::UpdateCity, game: &Game) -> anyhow::Result<()> {
        self.pos = data
            .pos
            .clone()
            .ok_or_else(|| anyhow!("missing city position"))?
            .into();
        self.owner = game.resolve_player_id(data.owner_id as u32)?;
        self.city_yield = data.r#yield.clone().unwrap_or_default().into();
        self.build_task = data
            .build_task
            .as_ref()
            .map(|b| BuildTask::from_data(b, game))
            .transpose()?;

        self.buildings = data
            .building_names
            .iter()
            .map(|name| game.registry().building(name))
            .collect::<Result<_, RegistryItemNotFound>>()?;
        self.resources = data
            .resources
            .iter()
            .map(|id| game.registry().resource(id))
            .collect::<Result<_, RegistryItemNotFound>>()?;

        if let Some(culture_values) = &data.culture_values {
            self.culture.set_data(game, culture_values)?;
        }

        self.data = data;

        Ok(())
    }

    pub fn id(&self) -> CityId {
        self.id
    }

    pub fn network_id(&self) -> u32 {
        self.data.id as u32
    }

    pub fn owner(&self) -> PlayerId {
        self.owner
    }

    pub fn city_yield(&self) -> Yield {
        self.city_yield
    }

    pub fn build_task(&self) -> Option<&BuildTask> {
        self.build_task.as_ref()
    }

    pub fn pos(&self) -> UVec2 {
        self.pos.clone()
    }

    pub fn num_culture(&self) -> i32 {
        self.data.culture
    }

    pub fn culture_needed(&self) -> i32 {
        self.data.culture_needed
    }

    pub fn buildings(&self) -> impl DoubleEndedIterator<Item = &Handle<Building>> + '_ {
        self.buildings.iter()
    }

    pub fn resources(&self) -> impl DoubleEndedIterator<Item = &Handle<Resource>> + '_ {
        self.resources.iter()
    }

    pub fn population(&self) -> i32 {
        self.data.population
    }

    pub fn stored_food(&self) -> i32 {
        self.data.stored_food
    }

    pub fn food_needed_for_growth(&self) -> i32 {
        self.data.food_needed_for_growth
    }

    pub fn consumed_food(&self) -> i32 {
        self.data.consumed_food
    }

    pub fn is_growing(&self) -> bool {
        self.consumed_food() < self.city_yield().food as i32
    }

    pub fn is_starving(&self) -> bool {
        self.consumed_food() > self.city_yield().food as i32
    }

    pub fn is_stagnant(&self) -> bool {
        self.consumed_food() == self.city_yield().food as i32
    }

    pub fn is_capital(&self) -> bool {
        self.data.is_capital
    }

    pub fn worked_tiles(&self) -> impl DoubleEndedIterator<Item = UVec2> + '_ {
        self.data.worked_tiles.iter().map(|p| p.clone().into())
    }

    pub fn manual_worked_tiles(&self) -> impl DoubleEndedIterator<Item = UVec2> + '_ {
        self.data
            .manual_worked_tiles
            .iter()
            .map(|p| p.clone().into())
    }

    pub fn is_tile_manually_worked(&self, tile: UVec2) -> bool {
        self.manual_worked_tiles().any(|t| t == tile)
    }

    pub fn num_happiness(&self) -> u32 {
        self.data.happiness_sources.iter().map(|s| s.count).sum()
    }

    pub fn num_health(&self) -> u32 {
        self.data.health_sources.iter().map(|s| s.count).sum()
    }

    pub fn num_unhappiness(&self) -> u32 {
        self.data.unhappiness_sources.iter().map(|s| s.count).sum()
    }

    pub fn num_sickness(&self) -> u32 {
        self.data.sickness_sources.iter().map(|s| s.count).sum()
    }

    pub fn happiness(&self) -> impl Iterator<Item = &HappinessEntry> {
        self.data.happiness_sources.iter()
    }

    pub fn unhappiness(&self) -> impl Iterator<Item = &UnhappinessEntry> {
        self.data.unhappiness_sources.iter()
    }

    pub fn health(&self) -> impl Iterator<Item = &HealthEntry> {
        self.data.health_sources.iter()
    }

    pub fn sickness(&self) -> impl Iterator<Item = &SicknessEntry> {
        self.data.sickness_sources.iter()
    }

    pub fn culture(&self) -> &Culture {
        &self.culture
    }

    pub fn culture_per_turn(&self) -> i32 {
        self.data.culture_per_turn
    }

    pub fn culture_level(&self) -> &str {
        &self.data.culture_level
    }

    pub fn beakers_per_turn(&self, game: &Game) -> u32 {
        (self.city_yield().commerce as f32 * game.the_player().beaker_percent() as f32 / 100.)
            .floor() as u32
    }

    pub fn gold_per_turn(&self, game: &Game) -> u32 {
        self.city_yield().commerce - self.beakers_per_turn(game)
    }

    pub fn culture_defense_bonus(&self) -> i32 {
        self.data.culture_defense_bonus
    }

    pub fn estimate_build_time_for_task(&self, task: &BuildTask) -> u32 {
        (task.cost - task.progress + self.city_yield().hammers - 1) / (self.city_yield().hammers)
    }

    pub fn estimate_remaining_build_time(&self) -> u32 {
        match &self.build_task {
            Some(task) => self.estimate_build_time_for_task(task),
            None => 0,
        }
    }

    pub fn turns_needed_for_growth(&self) -> u32 {
        (self.food_needed_for_growth() as u32 - self.stored_food() as u32 + self.city_yield().food
            - self.consumed_food() as u32
            - 1)
            / (self.city_yield().food - self.consumed_food() as u32)
    }

    pub fn maintenance_cost(&self) -> i32 {
        self.data.maintenance_cost
    }

    pub fn name(&self) -> &str {
        &self.data.name
    }

    pub fn set_previous_build_task(&mut self, task: PreviousBuildTask) {
        self.previous_build_task = Some(task);
    }

    pub fn previous_build_task(&self) -> Option<&PreviousBuildTask> {
        self.previous_build_task.as_ref()
    }
}

/// Something a city is building.
#[derive(Debug)]
pub struct BuildTask {
    pub cost: u32,
    pub progress: u32,
    pub kind: BuildTaskKind,
}

impl BuildTask {
    pub fn from_data(data: &protocol::BuildTask, game: &Game) -> anyhow::Result<Self> {
        let kind = match &data.kind {
            Some(k) => match &k.task {
                Some(protocol::build_task_kind::Task::Unit(t)) => {
                    BuildTaskKind::Unit(game.registry().unit_kind(&t.unit_kind_id)?)
                }
                Some(protocol::build_task_kind::Task::Building(t)) => {
                    BuildTaskKind::Building(game.registry().building(&t.building_name)?)
                }
                None => bail!("missing build task kind"),
            },
            None => bail!("missing build task kind"),
        };

        Ok(Self {
            cost: data.cost.try_into()?,
            progress: data.progress.try_into()?,
            kind,
        })
    }

    pub fn name(&self) -> &str {
        match &self.kind {
            BuildTaskKind::Unit(u) => &u.name,
            BuildTaskKind::Building(b) => &b.name,
        }
    }
}

#[derive(Debug)]
pub enum BuildTaskKind {
    Unit(Handle<UnitKind>),
    Building(Handle<Building>),
}

/// The previous build task completed by a city.
#[derive(Debug)]
pub struct PreviousBuildTask {
    pub task: BuildTask,
    /// Whether the task completed successfully.
    /// If false, then it was canceled, e.g. because
    /// we lost the necessary resources.
    pub succeeded: bool,
}
